demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -307,18 +310,23 @@ static bool matchTopic( const char * pTopicName,
         status = ( strncmp( pTopicName, pTopicFilter, topicNameLength ) == 0 ) ? true : false;
     }
 
-    /* Determine if topic filters starts with a wildcard. */
-    topicFilterStartsWithWildcard = ( ( pTopicFilter[ 0 ] == '+' ) ||
-                                      ( pTopicFilter[ 0 ] == '#' ) ) ? true : false;
-
-    /* If an exact match was not found, match against wildcard characters in topic filter.
-     * Note: According to the MQTT 3.1.1 specification, incoming PUBLISH topic names starting
-     * the "$" character cannot be matched against topic filter starting with a wildcard.
-     * i.e. for example, "$SYS/sport" cannot be matched with "#" or "+/sport" topic filters. */
-    if( ( status == false ) && !( ( pTopicName[ 0 ] == '$' ) &&
-                                  ( topicFilterStartsWithWildcard == true ) ) )
+    if( status == false )
     {
-        status = matchTopicFilter( pTopicName, topicNameLength, pTopicFilter, topicFilterLength );
+        /* If an exact match was not found, match against wildcard characters in
+         * topic filter.*/
+
+        /* Determine if topic filter starts with a wildcard. */
+        topicFilterStartsWithWildcard = ( ( pTopicFilter[ 0 ] == '+' ) ||
+                                          ( pTopicFilter[ 0 ] == '#' ) ) ? true : false;
+
+        /* Note: According to the MQTT 3.1.1 specification, incoming PUBLISH topic names
+         * starting the "$" character cannot be matched against topic filter starting with
+         * a wildcard, i.e. for example, "$SYS/sport" cannot be matched with "#" or
+         * "+/sport" topic filters. */
+        if( !( ( ( pTopicName[ 0 ] == '$' ) ) && ( topicFilterStartsWithWildcard == true ) ) )
```

Comment:
```suggestion
        if( !( ( pTopicName[ 0 ] == '$' ) && ( topicFilterStartsWithWildcard == true ) ) )
```
or even
```suggestion
        if( ( pTopicName[ 0 ] != '$' ) || ( topicFilterStartsWithWildcard == false ) )
```

Reply:
I would prefer the first of the two suggestions as it more clearly expresses the intention which is - do not match the topic starting with $ with filters starting with wildcards.